src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,495 @@
+    return $ ReorgVersion version
+
+  -- enforce invariant that the history has (B_restore-1,H_parent).
+  historyInvariant <- callDb "handleVersion" $ \db -> do
+        res <- qry db "SELECT COUNT(*) FROM BlockHistory WHERE blockheight = ? AND hash = ?;"
+            [SInt $ fromIntegral $ pred bRestore, SBlob (Data.Serialize.encode hsh)] [RInt]
+            >>= expectSingleRowCol "handleVersion: (historyInvariant):"
+        return res
+  when (historyInvariant /= SInt 1) $
+    internalError "handleVersion: History invariant violation"
+
+  case compare bRestore (bCurrent + 1) of
+   GT -> internalError "handleVersion: Block_Restore invariant violation!"
+   EQ -> do
+     assign bsBlockVersion (BlockVersion bRestore vCurrent)
+     (SInt txid) <- callDb "getting txid" $ \db ->
+       expectSingleRowCol "blah" =<< qry db
+         "SELECT endingtxid FROM BlockHistory WHERE blockheight = ?;"
+         [SInt (fromIntegral bCurrent)]
+         [RInt]
+     bs@(BlockVersion bh version) <- gets _bsBlockVersion
+     result <- callDb "handleVersion" $ \db -> qry db
+         "SELECT txid FROM VersionHistory WHERE blockheight = ? AND version = ?;"
+         [SInt (fromIntegral bh), SInt (fromIntegral version)]
+         [RInt]
+     newtxid <- if null result
+       then do
+         versionHistoryInsert bs (fromIntegral txid)
+         return (fromIntegral txid :: TxId)
+       else do
+        r <- liftIO $ expectSingleRowCol "handleVersion: block version check" result
+        case r of
+          SInt res -> return (fromIntegral res)
+          _ -> error "lol"
+     assign bsTxId newtxid
+     return newtxid
+   LT -> do
+     bsBlockVersion .= BlockVersion bRestore (succ vCurrent)
+     tableMaintenanceRowsVersionedTables
+     dropUserTables
+     t <- deleteHistory
+     assign bsTxId t
+     b <- gets _bsBlockVersion
+     versionHistoryInsert b t
+     return t
+
+tableMaintenanceRowsVersionedTables :: BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedTables = do
+  tableMaintenanceRowsVersionedSystemTables
+  tableMaintenanceRowsVersionedUserTables
+
+tableMaintenanceRowsVersionedSystemTables :: BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedSystemTables  = do
+  (BlockVersion bh _) <- gets _bsBlockVersion
+  callDb "tableMaintenanceRowsVersionedSystemTables" $ \db -> do
+    exec' db "DELETE FROM [SYS:keysets] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+    exec' db "DELETE FROM [SYS:modules] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+    exec' db "DELETE FROM [SYS:namespaces] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+    exec' db "DELETE FROM [SYS:pacts] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
+
+tableMaintenanceRowsVersionedUserTables :: BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedUserTables = do
+  (BlockVersion bh _) <- gets _bsBlockVersion
+  tblNames <- callDb "tableMaintenanceRowsVersionedUserTables" $ \db -> qry db
```

Comment:
Can we float the `callDb` out so that there's one `callDb` and everything happens under one lambda? Repeatedly using ` callDb` is going to waste time pointlessly twiddling the mutex

Reply:
Sure can do.